File: Problem234.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head.next is None:
            return True
        tail = self.findMid(head)
        midHead = tail.next
        tail.next = None
        revMidHead = self.reverseList(midHead)
        result = self.compare2Lists(head, revMidHead)
        midHead = self.reverseList(revMidHead)
        tail.next = midHead
        return result
        
    def findMid(self, head) -> ListNode:
        pre = head
        walk = head.next
        run = head.next.next
        while run is not None and run.next is not None:
            pre = pre.next
            walk = walk.next
            run = run.next.next
        return pre
        
    def reverseList(self, head) -> ListNode:
        pre = None
        curr = head
        while curr is not None:
            temp = curr.next
            curr.next = pre
            pre = curr
            curr = temp
        return pre
        
    def compare2Lists(self, head1, head2) -> bool:
        while head1 is not None and head2 is not None:
            if head1.val != head2.val:
                return False
            head1 = head1.next
            head2 = head2.next
        return True

File: Problem234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode *mid = findMid(head);
        ListNode *revHead = reverseList(mid->next);
        ListNode *l = head;
        ListNode *r = revHead;
        bool isPal = true;
        while (isPal && r) {
            if (l->val != r->val) {
                isPal = false;
            }
            l = l->next;
            r = r->next;
        }
        return isPal;
    }

private:
    ListNode *reverseList(ListNode* head) {
        ListNode *prevNode = nullptr;
        ListNode *currNode = head;
        ListNode *nextNode = head ? head->next : head;
        while (currNode) {
            currNode->next = prevNode;
            prevNode = currNode;
            currNode = nextNode;
            nextNode = nextNode ? nextNode->next : nextNode;
        }
        return prevNode;
    }

    ListNode *findMid(ListNode* head) {
        ListNode *walk = head;
        ListNode *run = head;
        while (run->next && run->next->next) {
            run = run->next->next;
            walk = walk->next;
        }
        return walk;
    }
};

// linked list
// time: O(n)
// space: O(1)

